examples/plugins/sample_pcm_plugin.c

```diff
@@ -0,0 +1,331 @@
+#include <limits.h>
+#include <sys/time.h>
+#include <sys/mman.h>
+#include <sound/asound.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <tinyalsa/plugin.h>
+#include <tinyalsa/asoundlib.h>
+
+/* 2 words of uint32_t = 64 bits of mask */
+#define PCM_MASK_SIZE (2)
+#define PCM_FORMAT_BIT(x) (1ULL << x)
+
+struct sample_pcm_priv {
+    FILE *fptr;
+    int session_id;
+    int channels;
+    int bitwidth;
+    int sample_rate;
+    unsigned int period_size;
+    snd_pcm_uframes_t total_size_frames;
+};
+
+struct pcm_plugin_hw_constraints sample_pcm_constrs = {
+    .access = 0,
+    .format = 0,
+    .bit_width = {
+        .min = 16,
+        .max = 32,
+    },
+    .channels = {
+        .min = 1,
+        .max = 8,
+    },
+    .rate = {
+        .min = 8000,
+        .max = 384000,
+    },
+    .periods = {
+        .min = 1,
+        .max = 8,
+    },
+    .period_bytes = {
+        .min = 96,
+        .max = 122880,
+    },
+};
+
+static inline struct snd_interval *param_to_interval(struct snd_pcm_hw_params *p,
+                                                  int n)
+{
+    return &(p->intervals[n - SNDRV_PCM_HW_PARAM_FIRST_INTERVAL]);
+}
+
+static inline int param_is_interval(int p)
+{
+    return (p >= SNDRV_PCM_HW_PARAM_FIRST_INTERVAL) &&
+        (p <= SNDRV_PCM_HW_PARAM_LAST_INTERVAL);
+}
+
+static unsigned int param_get_int(struct snd_pcm_hw_params *p, int n)
+{
+    if (param_is_interval(n)) {
+        struct snd_interval *i = param_to_interval(p, n);
+        if (i->integer)
+            return i->max;
+    }
+    return 0;
+}
+
+static inline struct snd_mask *param_to_mask(struct snd_pcm_hw_params *p, int n)
+{
+    return &(p->masks[n - SNDRV_PCM_HW_PARAM_FIRST_MASK]);
+}
+
+static inline int param_is_mask(int p)
+{
+    return (p >= SNDRV_PCM_HW_PARAM_FIRST_MASK) &&
+        (p <= SNDRV_PCM_HW_PARAM_LAST_MASK);
+}
+
+static inline int snd_mask_val(const struct snd_mask *mask)
+{
+    int i;
+    for (i = 0; i < PCM_MASK_SIZE; i++) {
+        if (mask->bits[i])
+            return ffs(mask->bits[i]) + (i << 5) - 1;
+    }
+    return 0;
+}
+
+static int alsaformat_to_bitwidth(int format)
+{
+    switch (format) {
+    case SNDRV_PCM_FORMAT_S32_LE:
+    case SNDRV_PCM_FORMAT_S24_LE:
+        return 32;
+    case SNDRV_PCM_FORMAT_S8:
+        return 8;
+    case SNDRV_PCM_FORMAT_S24_3LE:
+        return 24;
+    default:
+    case SNDRV_PCM_FORMAT_S16_LE:
+        return 16;
+    };
+}
+
+static int param_get_mask_val(struct snd_pcm_hw_params *p,
+                                        int n)
+{
+    if (param_is_mask(n)) {
+        struct snd_mask *m = param_to_mask(p, n);
+        int val = snd_mask_val(m);
+
+        return alsaformat_to_bitwidth(val);
+    }
+    return 0;
+}
+
+static int sample_session_open(int sess_id, unsigned int mode, struct sample_pcm_priv *priv)
+{
+     char fname[128];
+
+     snprintf(fname, 128, "/data/test_%d.raw", sess_id);
```

Comment:
Perhaps there's something I don't know here, but does this have to be in `/data/`? I feel like it would be better to just use the current directory, with a pattern like `sample_pcm_data_%.raw`. That way, the user does not have to potentially escalate their privileges.

Reply:
sure